hadoop-hdfs-project/hadoop-hdfs/src/test/java/org/apache/hadoop/hdfs/server/namenode/TestDecommissioningStatus.java

```diff
@@ -383,30 +383,70 @@ public void testDecommissionStatusAfterDNRestart() throws Exception {
 
   /**
    * Verify the support for decommissioning a datanode that is already dead.
-   * Under this scenario the datanode should immediately be marked as
-   * DECOMMISSIONED
+   * Under this scenario the datanode should be marked as
+   * DECOMMISSION_IN_PROGRESS first. When pendingReplicationBlocksCount and
+   * underReplicatedBlocksCount are both 0, it becomes DECOMMISSIONED.
    */
   @Test(timeout=120000)
   public void testDecommissionDeadDN() throws Exception {
     Logger log = Logger.getLogger(DatanodeAdminManager.class);
     log.setLevel(Level.DEBUG);
-    DatanodeID dnID = cluster.getDataNodes().get(0).getDatanodeId();
-    String dnName = dnID.getXferAddr();
-    DataNodeProperties stoppedDN = cluster.stopDataNode(0);
-    DFSTestUtil.waitForDatanodeState(cluster, dnID.getDatanodeUuid(),
-        false, 30000);
+
+    DistributedFileSystem fileSystem = cluster.getFileSystem();
+
+    // Create a file with one block. That block has one replica.
+    Path f = new Path("decommission.dat");
+    DFSTestUtil.createFile(fileSystem, f, fileSize, fileSize, fileSize,
+        (short)1, seed);
+
+    // Find the DN that owns the only replica.
+    RemoteIterator<LocatedFileStatus> fileList =
+        fileSystem.listLocatedStatus(f);
+    BlockLocation[] blockLocations = fileList.next().getBlockLocations();
+    String[] dnNames = blockLocations[0].getNames();
+
+    // Stop the DN leads to 1 block under-replicated
+    DataNodeProperties[] stoppedDNs = new DataNodeProperties[dnNames.length];
+    for (int i = 0; i < dnNames.length; i++) {
+      stoppedDNs[i] = cluster.stopDataNode(dnNames[i]);
+    }
+
     FSNamesystem fsn = cluster.getNamesystem();
     final DatanodeManager dm = fsn.getBlockManager().getDatanodeManager();
-    DatanodeDescriptor dnDescriptor = dm.getDatanode(dnID);
-    decommissionNode(dnName);
+    final List<DatanodeDescriptor> dead = new ArrayList<DatanodeDescriptor>();
+    while (true) {
+      dm.fetchDatanodes(null, dead, false);
+      if (dead.size() == 3) {
```

Comment:
Why waiting for `dead.size()==3`? They all seem to be the same host.

And it would be better to use `GenericTestUtils.waitFor` instead of using the `while(true)` loop.